x-pack/plugin/text-structure/src/main/java/org/elasticsearch/xpack/textstructure/transport/TransportFindFieldStructureAction.java

```diff
@@ -0,0 +1,214 @@
+        this.securityContext = XPackSettings.SECURITY_ENABLED.get(settings)
+            ? new SecurityContext(settings, threadPool.getThreadContext())
+            : null;
+    }
+
+    @Override
+    protected void doExecute(Task task, FindFieldStructureAction.Request request, ActionListener<TextStructureResponse> listener) {
+
+        if (licenseState.isSecurityEnabled()) {
+            final String[] indices = request.getIndices();
+
+            final String username = securityContext.getUser().principal();
+            final HasPrivilegesRequest privRequest = new HasPrivilegesRequest();
+            privRequest.applicationPrivileges(new RoleDescriptor.ApplicationResourcePrivileges[0]);
+            privRequest.username(username);
+            privRequest.clusterPrivileges(Strings.EMPTY_ARRAY);
+            privRequest.indexPrivileges(RoleDescriptor.IndicesPrivileges.builder().indices(indices).privileges(SearchAction.NAME).build());
+
+            ActionListener<HasPrivilegesResponse> privResponseListener = ActionListener.wrap(
+                r -> handlePrivsResponse(username, request, r, listener),
+                listener::onFailure
+            );
+            client.execute(HasPrivilegesAction.INSTANCE, privRequest, privResponseListener);
+
+        } else {
+            searchAndBuildTextStructureResponse(request, threadPool.getThreadContext().getHeaders(), listener);
+        }
+
+    }
+
+    private void handlePrivsResponse(
+        String username,
+        FindFieldStructureAction.Request request,
+        HasPrivilegesResponse response,
+        ActionListener<TextStructureResponse> listener
+    ) throws IOException {
+        if (response.isCompleteMatch()) {
+            searchAndBuildTextStructureResponse(request, threadPool.getThreadContext().getHeaders(), listener);
+        } else {
+            XContentBuilder builder = JsonXContent.contentBuilder();
+            builder.startObject();
+            for (ResourcePrivileges index : response.getIndexPrivileges()) {
+                builder.field(index.getResource());
+                builder.map(index.getPrivileges());
+            }
+            builder.endObject();
+
+            listener.onFailure(
+                Exceptions.authorizationError(
+                    "Cannot determine field structure because user {} lacks permissions on the indices: {}",
+                    username,
+                    Strings.toString(builder)
+                )
+            );
+        }
+    }
+
+    private void searchAndBuildTextStructureResponse(
+        FindFieldStructureAction.Request request,
+        Map<String, String> headers,
+        ActionListener<TextStructureResponse> listener
+    ) {
+
+        ActionListener<SearchResponse> searchResponseActionListener = ActionListener.wrap(searchResponse -> {
+            // TODO could we just feed the doc fields as "lines" to the field structure finder?
+            StringBuilder stringBuilder = new StringBuilder();
```

Comment:
I think it might be way more effecient to make better assumptions if the data is already inside elasticsearch. Like, defaulting to no merged lines (and if merged lines == 0 then simply feed the lines directly to the extractor). 

Also, we know that there will be no "header_row" since document order is not guaranteed. Should we even worry about supporting anything other than Semi-structured in this???